Add ListGroups method to Retrieve groups user belongs to

kars7e opened the pull request:
Adds a method to v3 bindings to retrieve groups user belongs to,
based on his/her ID.

For #382 

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

groups resource definition in keystone request router: https://github.com/openstack/keystone/blob/a3aee6ccb52d85eac1deedec31724a955d47fa96/keystone/identity/routers.py#L77-L84

list_group_for_users general implementation https://github.com/openstack/keystone/blob/16f6ed14df136eb9c283a5415b9f18fcb0834350/keystone/identity/core.py#L1281

list_group_for_users SQL driver implementation https://github.com/openstack/keystone/blob/7754f170aa0eb42bea356e7f912bc241832eb1f3/keystone/identity/backends/sql.py#L312


Based on: https://developer.openstack.org/api-ref/identity/v3/?expanded=list-groups-to-which-a-user-belongs-detail,list-users-detail#list-groups-to-which-a-user-belongs

jrperritt commented:
@kars7e 
> Links to the line numbers/files in the OpenStack source code

You've provided a link to the docs; that's not the same and not sufficient. This won't be reviewed until the source code links are provided.

kars7e commented:
Thanks @jrperritt and sorry for missing this requirement, I have updated the PR description.

jtopjian commented:
@kars7e Be careful when working with the Identity API. Rabbits love to dig there 😄 

A few comments/heads-up:

* Check out all of the resources I've collected in https://github.com/gophercloud/gophercloud/issues/287
* Also checkout the recent merge of Users Get/List: https://github.com/gophercloud/gophercloud/pull/288

It looks like the `Group` struct you made would be the exact same as what would go into a `Group` package. Can you confirm that? If so, I wonder if it's possible to define the `Group` result struct in a `groups` package and have it extracted in the `users.ListGroups` function. This would cut down on duplicated code.

If the groups returned in the `users.ListGroups` result is the same structure as what would be in a call to `groups.Get`, then note the following below:

* Note the `Extra` field in the above. It looks like Groups support extra attributes, too (see why we can't trust the API documentation? 😄 ):

```sql
MariaDB [keystone]> describe `group`;
+-------------+-------------+------+-----+---------+-------+
| Field       | Type        | Null | Key | Default | Extra |
+-------------+-------------+------+-----+---------+-------+
| id          | varchar(64) | NO   | PRI | NULL    |       |
| domain_id   | varchar(64) | NO   | MUL | NULL    |       |
| name        | varchar(64) | NO   |     | NULL    |       |
| description | text        | YES  |     | NULL    |       |
| extra       | text        | YES  |     | NULL    |       |
+-------------+-------------+------+-----+---------+-------+
5 rows in set (0.00 sec)
```

Now, you can't add `extra` data through the command-line, but you can do:

```sql
update `group` set extra = '{"hello":"world"}' where id = '7cf1576e744844c6886af09e495e62ba';
```

And in a response you get:

```json
{"links": 
  {"self": "http://127.0.0.1:35357/v3/groups?name=foo", "previous": null, "next": null}, 
  "groups": [
    {"description": "", 
    "links": {"self": "http://127.0.0.1:35357/v3/groups/7cf1576e744844c6886af09e495e62ba"}, 
    "hello": "world", 
    "id": "7cf1576e744844c6886af09e495e62ba", 
    "domain_id": 
    "default", 
    "name": "foo"
  }]
}
```

* Also be mindful of an `Options` field, too. Groups might support that one, as well.

jtopjian commented:
From playing with the API, it does look like what is returned in `/users/<id>/groups` is the same as `/groups/<id>`:

```shell
$ curl -s -H "X-Auth-Token: $token" http://127.0.0.1:35357/v3/users/dc2a24ccdeb046b988c52f1a6d8cf608/groups | python -mjson.tool
```

```json
{
    "groups": [
        {
            "description": "",
            "domain_id": "default",
            "hello": "world",
            "id": "7cf1576e744844c6886af09e495e62ba",
            "links": {
                "self": "http://127.0.0.1:35357/v3/groups/7cf1576e744844c6886af09e495e62ba"
            },
            "name": "foo"
        }
    ],
    "links": {
        "next": null,
        "previous": null,
        "self": "http://127.0.0.1:35357/v3/users/dc2a24ccdeb046b988c52f1a6d8cf608/groups"
    }
}
```

```shell
$ curl -s -H "X-Auth-Token: $token" http://127.0.0.1:35357/v3/groups/7cf1576e744844c6886af09e495e62ba
```

```json
{
    "group": {
        "description": "",
        "domain_id": "default",
        "hello": "world",
        "id": "7cf1576e744844c6886af09e495e62ba",
        "links": {
            "self": "http://127.0.0.1:35357/v3/groups/7cf1576e744844c6886af09e495e62ba"
        },
        "name": "foo"
    }
}
```

I was thinking only a subset of fields would be returned with the `/users/<id>/groups` call, ignoring the extra fields, but it looks like they're all there.

jrperritt commented:
@jtopjian All good points. One thing to think about here would be the import that you suggested: if we have `users` importing `groups` and (presumably soon) `groups` importing `users`, we'll get an import cycle, and the code won't compile.

kars7e commented:
Thanks @jtopjian for a thorough review. I've updated PR with support for `extra` in groups, I also added an acceptance test. 
WRT `options`, based on the content of this: https://github.com/openstack/keystone/blob/master/keystone/identity/backends/resource_options.py I understand `options` currently exists only for `users`. 
And finally, the `Group` type. I like the idea of defining this in `groups`, but it doesn't exist yet. If I was to create it, then it would make sense to add some actual methods to this package as well. Eventually, we get to the point where we get cyclic imports, as @jrperritt noted. I'm thinking how to solve that. Should we define some separate package for shared types?

jtopjian commented:
> I understand options currently exists only for users.

:+1: 

Thank you for looking into that.

> Should we define some separate package for shared types?

That was my thought as well. The shared package (`v3/shared`?) would contain the `User` and `Group` structs with their `UnmarshalJSON` method.

The other option is to just have duplicates in both the `users` and `groups` packages.

Unless there's a third option, my vote is for the shared package.

> Eventually, we get to the point where we get cyclic imports

This might be sooner than we think, especially if @mfranc213 follows up with the `groups.ListUsers` functionality.

And I can confirm that `groups.ListUsers` does return a full-fledged user and not a subset of user fields.

We should probably just figure out the best long-term structure now. @jrperritt ?

jtopjian commented:
Just a quick thought: would we be able to avoid the shared package as well as cyclic dependencies if the functions were placed in the opposite packages?

* `groups.ListGroupsOfUser(client, userid)`
* `users.ListUsersInGroup(client, groupid)`

The downside is that it's unintuitive.

jrperritt commented:
Alternatively, both functions could be in the `users` package:
`users.ListGroups(client, userid)`
`users.ListByGroup(client, groupid)`

jtopjian commented:
That's a good point. Since you suggested it, I'm in favor of it just to start being decisive. :)

But hopefully not delving into bike-shedding, `users.ListByGroup` gave me the immediate impression of "list all users, by group", implying more than just a single group membership would be returned.

What about: `users.ListInGroup`

Or more explicit: `users.ListMembersOfGroup`, though `Members` possibly adds a confusing synonym here.

jrperritt commented:
> implying more than just a single group membership would be returned

Understandable. I think the `groupid` parameter remedies that, but I think I'd be fine with `ListInGroup` instead.

jtopjian commented:
@kars7e Just a quick ping on this. It looks like this will need a rebase from other merges that have happened. Also, with regard to the discussion, let's just add everything to the `users` package.

kars7e commented:
Thanks @jtopjian for reminding me, I rebased the patch. Should I add the `ListInGroup` method in the same PR, or create a new one for it?

jtopjian commented:
@kars7e Thanks! It'd be best to open a new PR for `ListInGroup`.

jtopjian commented:
> let's just add everything to the users package.

This is my mistake and I apologize. I meant the two functions/methods being discussed would go in the users package, but everything else group-related (the result struct) should go in a `groups` package. This is so when someone implements CRUD functionality on groups, that information wouldn't be duplicated.

jtopjian commented on acceptance/openstack/identity/v3/users_test.go:
```diff
@@ -119,5 +119,37 @@ func TestUserCRUD(t *testing.T) {
+		t.Fatalf("Unable to list users: %v", err)
+	}
 
+	allUsers, err := users.ExtractUsers(allPages)
```
`allUserPages`

jtopjian replied:
Sorry, what I meant was:

```go
allUsers, err := users.ExtractUsers(allUserPages)
```

kars7e replied:
Wow, I really need to stop reviewing code at 1AM. Sorry, fixed!

jtopjian replied:
haha - no problem at all.

kars7e commented:
@jtopjian moved groups related things to groups package, take a look when you have some time

jtopjian commented:
This looks good to me.

Disclaimer: I did not do an in-depth review of the Groups API since we're technically only concerned about a group structure. I validated the fields, links, and result bodies from the Keystone schema and from interaction against a Keystone service. Everything looks good to me.

@jrperritt Thoughts / Comments on this PR?
